gobblin-iceberg/src/main/java/org/apache/gobblin/iceberg/writer/IcebergMetadataWriter.java

```diff
@@ -0,0 +1,828 @@
+    if (!tableMetadata.table.isPresent()) {
+      tableMetadata.table = Optional.of(catalog.loadTable(tid));
+    }
+    return tableMetadata.table.get();
+  }
+
+  /**
+   *  The method is used to get current watermark of the gmce topic partition for a table
+   *  Make the watermark config name contains topicPartition in case we change the gmce topic name for some reason
+   */
+  private Long getCurrentWaterMark(TableIdentifier tid, String topicPartition) {
+    if (tableCurrentWaterMarkMap.containsKey(tid)) {
+      return tableCurrentWaterMarkMap.get(tid);
+    }
+    org.apache.iceberg.Table icebergTable;
+    Long currentWatermark = DEFAULT_WATERMARK;
+    try {
+      icebergTable = getIcebergTable(tid);
+    } catch (NoSuchTableException e) {
+      return currentWatermark;
+    }
+    currentWatermark =
+        icebergTable.properties().containsKey(String.format(GMCE_HIGH_WATERMARK_KEY, topicPartition)) ? Long.parseLong(
+            icebergTable.properties().get(String.format(GMCE_HIGH_WATERMARK_KEY, topicPartition))) : DEFAULT_WATERMARK;
+    if (currentWatermark != DEFAULT_WATERMARK) {
+      // set the low watermark for current snapshot
+      tableMetadataMap.computeIfAbsent(tid, t -> new TableMetadata()).lowWatermark = Optional.of(currentWatermark);
+    }
+    return currentWatermark;
+  }
+
+  /**
+   *The write method will be responsible for process a given gmce and aggregate the metadata
+   * The logic of this function will be:
+   * 1. Check whether a table exists, if not then create the iceberg table
+   * 2. Compute schema from the gmce and update the cache for candidate schemas
+   * 3. Do the required operation of the gmce, i.e. addFile, rewriteFile or dropFile
+   * Note: this method only aggregate the metadata in cache without doing commitment. The actual commit will be done in flush method
+   */
+  @SuppressWarnings("checkstyle:WhitespaceAround")
+  public void write(GobblinMetadataChangeEvent gmce, Map<String, Collection<HiveSpec>> newSpecsMap,
+      Map<String, Collection<HiveSpec>> oldSpecsMap, HiveSpec tableSpec) throws IOException {
+    System.out.println("$$here");
+    TableIdentifier tid = TableIdentifier.of(tableSpec.getTable().getDbName(), tableSpec.getTable().getTableName());
+    TableMetadata tableMetadata = tableMetadataMap.computeIfAbsent(tid, t -> new TableMetadata());
+    Table table;
+    try {
+      table = getIcebergTable(tid);
+    } catch (NoSuchTableException e) {
+      try {
+        if (gmce.getOperationType() == OperationType.drop_files) {
+          log.warn("Table {} does not exist, skip processing this drop_file event", tid.toString());
+          return;
+        }
+        table = createTable(gmce, tableSpec);
+        tableMetadata.table = Optional.of(table);
+      } catch (Exception e1) {
+        log.error("skip processing {} for table {}.{} due to error when creating table", gmce.toString(),
```

Comment:
Normally the error is about union thing which iceberg does not support. I print e1 as debug level to avoid flushing the log since there will be many gmce received for the same table.